### PR TITLE
refactor: unify asset edit options 

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- The edit options for the bottom sheet were using different component for assets, that is why during the design review we found out them to have a wrong items for ephemeral messages

### Solutions

- Make use of the same component instead of two to do not repeat this error in the future.

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)
image

image
----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/4/) **failed**.

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/5/) **failed**.

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/6/) **failed**.

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/7/) **failed**.

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/8/) **failed**.

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/9/) **failed**.

**Comment by AndroidBob** [comment]
Build [10](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/10/) **failed**.

**Comment by AndroidBob** [comment]
Build [11](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/11/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt [hunk 1]
```diff
@@ -22,29 +22,30 @@ import com.wire.android.ui.edit.DeleteItemMenuOption
...
     return buildList {
-        if (isAvailable) {
+        if (!isUploading) {
+            if (!isEphemeral) add { ReactionOption(onReactionClick) }
             add { MessageDetailsMenuOption(onDetailsClick) }
-            if (isAssetMessage) add { DownloadAssetExternallyOption(onDownloadAsset) }
-            if (isGenericAsset) add { OpenAssetExternallyOption(onOpenAsset) }
-            add { DeleteItemMenuOption(onDeleteMessage) }
+            if (!isEphemeral) add { ReplyMessageOption(onReplyClick) }
+            add { DownloadAssetExternallyOption(onDownloadAsset) }
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt:45]: Can we still download the asset even though it is ephemeral?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt [hunk 2]
```diff
@@ -22,29 +22,30 @@ import com.wire.android.ui.edit.DeleteItemMenuOption
...
+            if (!isEphemeral) add { ShareAssetMenuOption(onShareAsset) }
+            if (onOpenAsset != null && !isEphemeral) add { OpenAssetExternallyOption(onOpenAsset) }
         }
+        add { DeleteItemMenuOption(onDeleteClick) }
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt:49]: By having it outside the `if (!isUploading) { ..` we are then able to delete the asset while its uploading, right?

**Comment by AndroidBob** [comment]
Build [13](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/13/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/EditMessageMenuItems.kt [hunk 3]
```diff
@@ -118,48 +101,49 @@ fun EditMessageMenuItems(
...
-            }
-            add { DeleteItemMenuOption(onDeleteItemClick) }
-        }
+        TextMessageEditMenuItems(
+            isEphemeral = message.expirationStatus is ExpirationStatus.Expirable,
+            isUploading = message.isPending,
+            onDeleteClick = onDeleteItemClick,
+            onDetailsClick = onDetailsItemClick,
+            onReactionClick = onReactionItemClick,
+            onEditClick = if (message.isMyMessage) onEditItemClick else null,
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/EditMessageMenuItems.kt:138]: To edit the message shouldn't we also check for:
`message.isTextMessage` and `localFeatureVisibilityFlags.MessageEditIcon`?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/TextMessageMenuItems.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+): List<@Composable () -> Unit> {
+    return buildList {
+        if (!isUploading) {
+            if (!isEphemeral) add { ReactionOption(onReactionClick) }
+            add { MessageDetailsMenuOption(onDetailsClick) }
+            if (!isEphemeral) add { CopyItemMenuOption(onCopyClick) }
+            if (!isEphemeral) add { ReplyMessageOption(onReplyClick) }
+            if (!isEphemeral && onEditClick != null) add { EditMessageMenuOption(onEditClick) }
+        }
+        add { DeleteItemMenuOption(onDeleteClick) }
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/TextMessageMenuItems.kt:45]: Same as above:
By having it outside the if (!isUploading) { .. we are then able to delete the asset while its uploading, right?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt [hunk 5]
```diff
@@ -54,14 +54,16 @@ sealed class UIMessage(
...
         val isDeleted: Boolean = header.messageStatus == Deleted
         val sendingFailed: Boolean = header.messageStatus is MessageStatus.MessageSendFailureStatus
         val decryptionFailed: Boolean = header.messageStatus is DecryptionFailure
         val receivingFailed: Boolean = header.messageStatus == ReceiveFailure || decryptionFailed
         val isAvailable: Boolean = !isDeleted && !sendingFailed && !receivingFailed
         val isPending: Boolean = header.messageStatus.isPending
         val isMyMessage = source == MessageSource.Self
+        val isAssetMessage = messageContent is UIMessageContent.AssetMessage
+        || messageContent is UIMessageContent.ImageMessage
+        || messageContent is UIMessageContent.AudioAssetMessage
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt:61]: I don't know so I'm asking:

Should we also consider checking for `RestrictedAsset`?

**Review by alexandreferris — COMMENTED** [review]
Great work! I just have some comments

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt (hunk 2 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt:49]: correct

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt (hunk 1 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/AssetEditMenuItems.kt:45]: yes from what I know we can 🤔

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/EditMessageMenuItems.kt (hunk 3 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/EditMessageMenuItems.kt:138]: we do not need to check for textMessage anymore as we can only edit text messages and I created a seperated component for edit optiosn for a text message also the FeatureVisibityFlags were used some time ago where and MessagEEditIcon is set to true, so bassicly we can remove this check

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/TextMessageMenuItems.kt (hunk 4 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/edit/TextMessageMenuItems.kt:45]: yes

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt (hunk 5 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/model/UIMessage.kt:61]: We did not check for it before 🤔  so I do not think so ?

**Comment by AndroidBob** [comment]
Build [14](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/14/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/gallery/MediaGalleryViewModel.kt [hunk 6]
```diff
@@ -73,37 +75,39 @@ class MediaGalleryViewModel @Inject constructor(
     private val getImageData: GetMessageAssetUseCase,
     private val fileManager: FileManager,
     private val deleteMessage: DeleteMessageUseCase,
+    private val getMessageByIdUseCase: GetMessageByIdUseCase
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/gallery/MediaGalleryViewModel.kt]: where are we using this use case?

**Comment by AndroidBob** [comment]
Build [15](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/15/) **failed**.

**Comment by AndroidBob** [comment]
Build [16](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/16/) **failed**.

**Comment by AndroidBob** [comment]
Build [17](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/17/) **failed**.

**Comment by AndroidBob** [comment]
Build [18](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/18/) **failed**.

**Comment by AndroidBob** [comment]
Build [19](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/19/) **failed**.

**Review by MohamadJaara — APPROVED** [review]

**Review by alexandreferris — APPROVED** [review]

**Review by Garzas — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [20](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/20/) **failed**.

**Comment by AndroidBob** [comment]
Build [21](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/21/) **failed**.

**Comment by AndroidBob** [comment]
Build [23](https://10.10.124.134/job/wire-android-reloaded/job/PR-1819/23/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.2.0(6838650)(6838650)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.2.0(6838650)(6838650)-staging-compat.apk)

**Pull request merged**